### PR TITLE
Create Orleans Quickstart

### PR DESCRIPTION
**Pull request by alexwolfmsft**
## Summary

Describe your changes here.

Fixes #Issue_Number (if available)

Preview:

- [Quickstart: Build your first Orleans app with ASP.NET Core](https://review.learn.microsoft.com/en-us/dotnet/orleans/quickstarts/build-your-first-orleans-app?branch=pr-en-us-33075&tabs=visual-studio)

### CODE REVIEW

File: docs/orleans/quickstarts/build-your-first-orleans-app.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+var app = builder.Build();
+
+app.Run();
+```
+
+## Configure the silos
+
+[Silos](/dotnet/orleans/overview#what-are-silos) are responsible for storing grains and are another core building block of Orleans. A silo can contain one or more grains; a group of silos is known as a cluster. The cluster coordinates work between silos, allowing communication with grains as though they were all available in a single process. 
+
+At the top of the Program class, refactor the builder code to use Orleans. The following code uses a `SiloBuilder` class to create a localhost cluster with a silo that can store grains. The `SiloBuilder` also uses the `AddMemoryGrainStorage` to configure the Orleans silos to persist grains in memory. This scenario uses local resources for development, but a production app can be configured to use highly scalable clusters and storage using services like Azure Blob Storage.
```
Comment (IEvangelist) [inline docs/orleans/quickstarts/build-your-first-orleans-app.md:110]:
```suggestion
At the top of the Program class, refactor the builder code to use Orleans. The following code uses a <xref:Orleans.Hosting.ISiloBuilder> class to create a localhost cluster with a silo that can store grains. The `ISiloBuilder` also uses the `AddMemoryGrainStorage` to configure the Orleans silos to persist grains in memory. This scenario uses local resources for development, but a production app can be configured to use highly scalable clusters and storage using services like Azure Blob Storage.
```

Also, xref to `AddMemoryGrainStorage`.

File: docs/orleans/quickstarts/build-your-first-orleans-app.md [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+});
+
+var app = builder.Build();
+
+app.Run();
+```
+
+## Create the URL shortener grain
+
+[Grains](/dotnet/orleans/overview#what-are-grains) are the most essential primitives and building blocks of Orleans applications. A grain is a class that inherits from the Grain base class, which manages various internal behaviors and integration points with the Orleans framework. Grains should also implement one of the interfaces listed below to define the type of grain key identifier. Each of these interfaces defines a similar contract, but marks your class with a different data type for the identifier that Orleans uses to track the grain, such as a string or integer.
```
Comment (IEvangelist) [inline docs/orleans/quickstarts/build-your-first-orleans-app.md:128]: Try to use relative links to articles, where possible.

Comment (IEvangelist) [inline docs/orleans/quickstarts/build-your-first-orleans-app.md:128]:
```suggestion
[Grains](/dotnet/orleans/overview#what-are-grains) are the most essential primitives and building blocks of Orleans applications. A grain is a class that inherits from the `Grain` base class (or `IGrainBase` interface), which manages various internal behaviors and integration points with the Orleans framework. Grains should also implement one of the interfaces listed below to define the type of grain key identifier. Each of these interfaces defines a similar contract, but marks your class with a different data type for the identifier that Orleans uses to track the grain, such as a string or integer.
```

File: docs/orleans/quickstarts/build-your-first-orleans-app.md [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+- IGrainWithGuidKey
+- IGrainWithIntegerKey
+- IGrainWithStringKey
+- IGrainWithGuidCompoundKey
+- IGrainWithIntegerCompoundKey
```
Comment (IEvangelist) [inline docs/orleans/quickstarts/build-your-first-orleans-app.md:134]:
```suggestion
- `IGrainWithGuidKey`
- `IGrainWithIntegerKey`
- `IGrainWithStringKey`
- `IGrainWithGuidCompoundKey`
- `IGrainWithIntegerCompoundKey`
```

File: docs/orleans/quickstarts/build-your-first-orleans-app.md [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        Task<string> GetUrl();
+    }
+    ```
+
+1. Create a `UrlShortenerGrain` class using the following code. This class inherits from the `Grain` class provided by Orleans and implements the `IUrlShortenerGrain` interface you created. The class also uses the `IPersistentState` interface of Orleans to manage reading and writing state values for the URLs to the configured silo storage.
+
+    ```csharp
+    public class UrlShortenerGrain : Grain, IUrlShortenerGrain
+    {
+        private IPersistentState<KeyValuePair<string, string>> _cache;
```
Comment (IEvangelist) [inline docs/orleans/quickstarts/build-your-first-orleans-app.md:158]:
```suggestion
        private readonly IPersistentState<KeyValuePair<string, string>> _cache;
```

File: docs/orleans/quickstarts/build-your-first-orleans-app.md [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    ```
+
+## Create the endpoints
+
+Next, create two endpoints to utilize the Orleans grain and silo configurations:
+
+- A `/shorten/{*url}` endpoint to handle creating and storing a shortened version of the URL. The original, full URL is provided as a path parameter, and the shortened URL is returned to the user for later use.
+- A `/go/{*shortenedUrl}` endpoint to handle redirecting users to the original URL using the shortened URL that is supplied as a parameter.
+
+Append the following code to the end of the `Program.cs` file:
```
Comment (IEvangelist) [inline docs/orleans/quickstarts/build-your-first-orleans-app.md:189]:
```suggestion
Append the following code to the end of the _Program.cs_ file:
```

File: docs/orleans/quickstarts/build-your-first-orleans-app.md [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+    ```dotnetcli
+    dotnet run
+    ```
+
+    The app should launch in the browser and display the familiar `Hello world!` text.
+
+2) In the browser address bar, test the `shorten` endpoint by entering a URL path such as `{localhost}/shorten/https://microsoft.com`. The page should reload and provide a shortened URL. Copy the shortened URL to your clipboard.
+
+3) Paste the shortened URL into the address bar and press enter. The page should reload and redirect you to [https://microsoft.com](https://microsoft.com).
```
Comment (IEvangelist) [inline docs/orleans/quickstarts/build-your-first-orleans-app.md:236]:
```suggestion
3) Paste the shortened URL into the address bar and press enter. The page should reload and redirect you to [https://microsoft.com](https://microsoft.com).

```

**Review by IEvangelist — COMMENTED** [review]
This draft looks great thus far, here is a bit of initial feedback.

File: docs/orleans/toc.yml [hunk 7]
```diff
@@ -10,6 +10,10 @@ items:
         href: benefits.md
       - name: What's new
         href: whats-new-in-orleans.md
+  - name: Get started
+    items:
+      - name: Build your first Orleans app
+        href: quickstarts/build-your-first-orleans-app.md
```
Comment (IEvangelist) [inline docs/orleans/toc.yml:16]:
```suggestion
      - name: Build your first Orleans app
        href: quickstarts/build-your-first-orleans-app.md
```

There is already a `Get started` node.

Comment (alexwolfmsft) [reply docs/orleans/toc.yml:16]: That was supposed to be Quickstarts, fixing

File: docs/orleans/quickstarts/build-your-first-orleans-app.md [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+        public UrlShortenerGrain(
+            [PersistentState(
+                stateName: "url",
+                storageName: "urls")]
+                IPersistentState<KeyValuePair<string, string>> state)
+        {
+            _cache = state;
+        }
+    
+        public async Task SetUrl(string shortenedRouteSegment, string fullUrl)
```
Comment (IEvangelist) [inline docs/orleans/quickstarts/build-your-first-orleans-app.md:168]:
```suggestion
        public async Task SetUrl(string fullUrl)
```

File: docs/orleans/quickstarts/build-your-first-orleans-app.md [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+
+- A `SetUrl` method to persist the original and shortened URLs.
+- A `GetUrl` method to retrieve the original URL using the shortened URL.
+
+1. Append the following interface definition to the bottom of the _Program.cs_ file.
+
+    ```csharp
+    public interface IUrlShortenerGrain : IGrainWithStringKey
+    {
+        Task SetUrl(string shortenedRouteSegment, string fullUrl);
```
Comment (IEvangelist) [inline docs/orleans/quickstarts/build-your-first-orleans-app.md:147]:
```suggestion
        Task SetUrl(string fullUrl);
```

File: docs/orleans/quickstarts/build-your-first-orleans-app.md [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+```csharp
+app.MapGet("/shorten/{*path}",
+    async (IGrainFactory grains, HttpRequest request, string path) =>
+    {
+        // Create a unique, short ID
+        var shortenedRouteSegment = Guid.NewGuid().GetHashCode().ToString("X");
+
+        // Create and persist a grain with the shortened ID and full URL
+        var shortenerGrain = grains.GetGrain<IUrlShortenerGrain>(shortenedRouteSegment);
+        await shortenerGrain.SetUrl(shortenedRouteSegment, path);
```
Comment (IEvangelist) [inline docs/orleans/quickstarts/build-your-first-orleans-app.md:199]:
```suggestion
        await shortenerGrain.SetUrl(path);
```

File: docs/orleans/quickstarts/build-your-first-orleans-app.md [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+var app = builder.Build();
+
+app.MapGet("/", () => "Hello World!");
+
+app.MapGet("/shorten/{*path}",
+    async (IGrainFactory grains, HttpRequest request, string path) =>
+    {
+        var shortenedRouteSegment = Guid.NewGuid().GetHashCode().ToString("X");
+        var shortenerGrain = grains.GetGrain<IUrlShortenerGrain>(shortenedRouteSegment);
+        await shortenerGrain.SetUrl(shortenedRouteSegment, path);
```
Comment (IEvangelist) [inline docs/orleans/quickstarts/build-your-first-orleans-app.md:246]:
```suggestion
        await shortenerGrain.SetUrl(path);
```

File: docs/orleans/quickstarts/build-your-first-orleans-app.md [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+        var url = await shortenerGrain.GetUrl();
+
+        return Results.Redirect(url);
+    });
+
+app.Run();
+
+public interface IUrlShortenerGrain : IGrainWithStringKey
+{
+    Task SetUrl(string shortenedRouteSegment, string fullUrl);
```
Comment (IEvangelist) [inline docs/orleans/quickstarts/build-your-first-orleans-app.md:268]:
```suggestion
    Task SetUrl(string fullUrl);
```

**Review by IEvangelist — APPROVED** [review]
This looks great, I have a few suggestions.

File: docs/orleans/quickstarts/build-your-first-orleans-app.md [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+                storageName: "urls")]
+                IPersistentState<KeyValuePair<string, string>> state)
+        {
+            _state = state;
+        }
+    
+        public async Task SetUrl(string fullUrl)
+        {
+            _state.State = new UrlDetails(this.GetPrimaryKeyString(), fullUrl);
+            await _cache.WriteStateAsync();
```
Comment (IEvangelist) [inline docs/orleans/quickstarts/build-your-first-orleans-app.md:171]:
```suggestion
            await _state.WriteStateAsync();
```

**Pull request merged**